rust/src/transformations/make_stable_expr/expr_discrete_quantile_score/plugin_dq_score.rs

```diff
@@ -0,0 +1,209 @@
+use std::iter::zip;
+
+use polars::{
+    datatypes::{
+        ArrayChunked, ArrowDataType, DataType::*, Field, Float32Type, Float64Type, Int16Type,
+        Int32Type, Int64Type, Int8Type, PolarsDataType, StaticArray, UInt32Type, UInt64Type,
+    },
+    error::{polars_bail, polars_err, PolarsResult},
+    series::Series,
+};
+use polars_arrow::{
+    array::{FixedSizeListArray, UInt64Array},
+    datatypes::Field as ArrowField,
+};
+use polars_plan::{
+    dsl::{GetOutput, SeriesUdf},
+    prelude::{ApplyOptions, FunctionOptions},
+};
+
+#[cfg(feature = "ffi")]
+use pyo3_polars::derive::polars_expr;
+#[cfg(feature = "ffi")]
+use serde::{Deserialize, Serialize};
+
+use crate::{core::OpenDPPlugin, error::Fallible, traits::RoundCast};
+
+/// Arguments for the discrete quantile score expression
+#[derive(Clone)]
+#[cfg_attr(feature = "ffi", derive(Serialize, Deserialize))]
+pub(crate) struct DQScoreArgs {
+    /// Candidates to score
+    pub candidates: Vec<f64>,
+    /// A value between [0, 1]
+    pub alpha: f64,
+    /// Alpha numerator, alpha denominator, and max partition length
+    pub constants: Option<(u64, u64, u64)>,
+}
+
+impl OpenDPPlugin for DQScoreArgs {
+    fn get_options(&self) -> FunctionOptions {
+        FunctionOptions {
+            collect_groups: ApplyOptions::GroupWise,
+            fmt_str: "dq_score",
+            returns_scalar: true,
+            changes_length: true,
+            ..Default::default()
+        }
+    }
+}
+
+// allow the LaplaceArgs struct to be stored inside an AnonymousFunction, when used from Rust directly
+impl SeriesUdf for DQScoreArgs {
+    // makes it possible to downcast the AnonymousFunction trait object back to LaplaceArgs
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+
+    fn call_udf(&self, s: &mut [Series]) -> PolarsResult<Option<Series>> {
+        dq_score_udf(s, self.clone()).map(Some)
+    }
+
+    fn get_output(&self) -> Option<GetOutput> {
+        let kwargs = self.clone();
+        Some(GetOutput::map_fields(move |fields| {
+            dq_score_type_udf(fields, kwargs.clone())
+                .ok()
+                .unwrap_or_else(|| fields[0].clone())
+        }))
+    }
+}
+
+/// Implementation of the Laplace noise expression.
+///
+/// The Polars engine executes this function over chunks of data.
+fn dq_score_udf(inputs: &[Series], kwargs: DQScoreArgs) -> PolarsResult<Series> {
+    let Ok([series]) = <&[_; 1]>::try_from(inputs) else {
+        polars_bail!(InvalidOperation: "Quantile expects a single input field");
+    };
+
+    let n = series.len() as u64;
+    let (candidates, constants) = (kwargs.candidates, kwargs.constants);
+    let Some((alpha_num, alpha_den, size_limit)) = constants else {
+        polars_bail!(InvalidOperation:
+            "encountered quantile in expression that has not been made private or stable",
```

Comment:
This would fail if not a triple? I'm not immediately seeing how that connects to the text of the error message... Might be more generic, or move this detailed message up to a higher level. 

But if the connection is clear to you, please ignore.

Reply:
Left a comment